### DISCUSSION TITLE
Initial conditions from a point cloud file

### DISCUSSION THREAD

**Discussion by makrandak** [Nek5000/nekRS#553] [Q&A]
Hello everyone, 

I am trying to set up a homogeneous isotropic turbulence case. I want to set the initial conditions from a model energy spectrum. I have a code that spits out coordinates and velocities in a point cloud. Is there a way to set the initial conditions from this file that contains x,y,z,vel_x, vel_y, and vel_z point cloud? I am assuming there would be a requirement to interpolate from the point cloud to the quadrature points.  If anyone has done this case and has a simpler way of doing this, I would be grateful to hear their suggestion as well. 

With sincere regards,
Makrand

**Comment by yslan** [comment]
In Nek, we have gslib/findpt that interpolates the solutions onto given (arbitrary) points. The other direction, which is a J^T from the existed interpolation matrix J, is not currently supported. There are people working on that though.

You can always go through all points, locate it to corresponding element, then do interpolation, sums the contribution from all points. But, this could be expansive depending on your implementation, especially, how the data binning is done.

I suggest to do it offline and test it with Nek5000 first. 

If you only need this once as the initial condition and the size is not large, I suggest to do this in MATLAB or python (You can find readers from KTH group https://github.com/eX-Mech/)
Basically, you will read a restart file containing mesh, sort your point cloud into each elements, construct the element-wise interpolation matrix from SEM basis to your points, apply transpose, write everything into a new restart file and finally restart Nek5000 or NekRS simulation from it.

Alternatively, if the points cloud can be represented by a finite elements mesh (Delaney triangulation, then converting into SEM mesh by tet2hex), then you can create a Nek restart file containing such arbitrary SEM mesh and call gfldr.

**Reply by clloydfluids** [reply]
Hello @yslan - Just found this thread as I am facing the same challenge as Makrand. Has there been any progress on this implementation in nekrs?

**Reply by yslan** [reply]
Nek has capability of interpolating solutions onto arbitrary grid. However, without implementing the discretization of such grid, it's not possible to interpolate the solution back. After all, J^T mentioned above, is not an interpolation.

What I'd suggest is, you first interpolate (says qubic spline) the solution on to a Nek mesh (e.g. a box mesh) in your application. Then, use gfldr to read the solution back into Nek.

**Comment by stgeke** [comment]
Why don't you use a linear forcing method (e.g. https://pubs.aip.org/aip/pof/article-abstract/28/3/035114/316071/Constant-energetics-physical-space-forcing-methods?redirectedFrom=fulltext) to generate a HIT field.

**Comment by makrandak** [comment]
@yslan Thank you very much for the clarification and the detailed answer. I decided to implement a function in the .udf file that evaluates the physical space velocities on the fly. Certainly not fast, but as it is only required at the initiation of the simulation, it is alright. 

@stgeke Yes, I am actually trying to implement linear forcing. Just wanted to try a sane initial condition instead of starting from a uniform velocity. I understand that the effect of the initial condition should vanish if the forcing kernel I have implemented is correct. 

I am new to nekRS and trying to modify things and I am getting used to writing the occa kernels for some interesting things I am doing to the standard NS. 

I appreciate your help and comments. 

With sincere regards,
Makrand

**Comment by ajariwala611** [comment]
Hello,

I have done something similar for setting PIV velocity fields as inlet inflow time-varying boundary condition by reading set of XYZ coordinates and corresponding velocities. I first read the files and bi-linearly interpolate them at each GLL points. You can find the GitHub repo [here](https://github.com/ajariwala611/NEKRS_v23/tree/main/x10_SJ_only).
